File: project/scratch/analyze_blade.php
<?php

/**
 * Blade File Connection Analyzer
 * Analyzes which blade files are connected to controllers (directly or via blade-to-blade)
 */

$projectRoot = __DIR__ . '/..';
$viewsDir    = $projectRoot . '/resources/views';
$controllersDir = $projectRoot . '/app/Http/Controllers';
$routesDir   = $projectRoot . '/routes';

// ─────────────────────────────────────────────
// 1. Collect all blade files (dot-path format)
// ─────────────────────────────────────────────
function getBladeFiles(string $dir, string $baseDir): array {
    $result = [];
    $rii = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($dir));
    foreach ($rii as $file) {
        if ($file->isFile() && str_ends_with($file->getFilename(), '.blade.php')) {
            $fullPath = $file->getRealPath();
            $relative = str_replace($baseDir . DIRECTORY_SEPARATOR, '', $fullPath);
            $relative = str_replace(DIRECTORY_SEPARATOR, '/', $relative);
            $dotPath  = str_replace('.blade.php', '', $relative);
            $dotPath  = str_replace('/', '.', $dotPath);
            $result[$dotPath] = $fullPath;
        }
    }
    ksort($result);
    return $result;
}

// ─────────────────────────────────────────────
// 2. Get all view() calls from Controllers + Routes
// ─────────────────────────────────────────────
function getViewCallsFromFiles(string $dir, string $pattern = '*.php'): array {
    $views = [];
    $files = [];

    // Recursive scan
    $rii = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($dir));
    foreach ($rii as $file) {
        if ($file->isFile() && str_ends_with($file->getFilename(), '.php')) {
            $files[] = $file->getRealPath();
        }
    }

    foreach ($files as $filepath) {
        $content = file_get_contents($filepath);
        // Match view('...') and view("...")
        preg_match_all('/\bview\s*\(\s*[\'"]([^\'"]+)[\'"]/i', $content, $matches);
        foreach ($matches[1] as $viewName) {
            $views[$viewName][] = $filepath;
        }
    }
    return $views;
}

// ─────────────────────────────────────────────
// 3. Get blade-to-blade connections (@include, @extends, @component)
// ─────────────────────────────────────────────
function getBladeToBlade(array $bladeFiles): array {
    $connections = []; // $parentDotPath => [child dot paths]
    foreach ($bladeFiles as $dotPath => $fullPath) {
        $content = file_get_contents($fullPath);
        $refs = [];

        // @extends('...')
        preg_match_all('/@extends\s*\(\s*[\'"]([^\'"]+)[\'"]\s*\)/i', $content, $m);
        foreach ($m[1] as $ref) $refs[] = $ref;

        // @include('...')
        preg_match_all('/@include(?:If|When|Unless|First)?\s*\(\s*[\'"]([^\'"]+)[\'"]/i', $content, $m);
        foreach ($m[1] as $ref) $refs[] = $ref;

        // @component('...')
        preg_match_all('/@component\s*\(\s*[\'"]([^\'"]+)[\'"]/i', $content, $m);
        foreach ($m[1] as $ref) $refs[] = $ref;

        if (!empty($refs)) {
            $connections[$dotPath] = array_unique($refs);
        }
    }
    return $connections;
}

// ─────────────────────────────────────────────
// MAIN ANALYSIS
// ─────────────────────────────────────────────

$allBlades = getBladeFiles($viewsDir, $viewsDir);

// Views referenced directly by Controllers
$controllerViewCalls = getViewCallsFromFiles($controllersDir);
// Views referenced in routes/web.php etc.
$routeViewCalls = getViewCallsFromFiles($routesDir);

// Merge all direct controller + route view calls
$allDirectViews = [];
foreach (array_merge($controllerViewCalls, $routeViewCalls) as $viewName => $files) {
    $allDirectViews[$viewName] = true;
}

// Blade-to-blade connections
$bladeToBlade = getBladeToBlade($allBlades);

// Build a reverse map: which blade files are @included/@extended BY another blade
$referencedByBlade = []; // dotPath => [parent dot paths]
foreach ($bladeToBlade as $parent => $children) {
    foreach ($children as $child) {
        $referencedByBlade[$child][] = $parent;
    }
}

// ─────────────────────────────────────────────
// Classify each blade file
// ─────────────────────────────────────────────
$directlyConnected    = []; // directly used in controller/route
$bladeConnected       = []; // used via blade-to-blade (not directly in controller)
$notConnected         = []; // not connected anywhere

foreach ($allBlades as $dotPath => $fullPath) {
    $isDirectController = isset($allDirectViews[$dotPath]);

    // Also check if any parent blade that uses this is connected to a controller
    $connectedViaBladeChain = false;
    $bladeParents = [];

    if (isset($referencedByBlade[$dotPath])) {
        foreach ($referencedByBlade[$dotPath] as $parent) {
            $bladeParents[] = $parent;
            if (isset($allDirectViews[$parent])) {
                $connectedViaBladeChain = true;
            }
        }
    }

    if ($isDirectController) {
        $directlyConnected[$dotPath] = [
            'path' => $fullPath,
            'referenced_in' => array_map(function($f) use ($projectRoot) {
                return str_replace($projectRoot . '/', '', $f);
            }, array_unique(array_merge(
                $controllerViewCalls[$dotPath] ?? [],
                $routeViewCalls[$dotPath] ?? []
            )))
        ];
    } elseif ($connectedViaBladeChain) {
        $bladeConnected[$dotPath] = [
            'path' => $fullPath,
            'included_by' => $bladeParents,
        ];
    } else {
        $notConnected[$dotPath] = [
            'path' => $fullPath,
            'included_by_blade' => $bladeParents,
        ];
    }
}

// ─────────────────────────────────────────────
// OUTPUT
// ─────────────────────────────────────────────

$divider = str_repeat('─', 80);

echo "\n";
echo "╔══════════════════════════════════════════════════════════════════════════════╗\n";
echo "║           BLADE FILE CONNECTION ANALYSIS REPORT                            ║\n";
echo "╚══════════════════════════════════════════════════════════════════════════════╝\n\n";
echo "Total blade files found: " . count($allBlades) . "\n\n";

// ── SECTION 1: Directly connected to controller ──
echo $divider . "\n";
echo "✅  DIRECTLY CONNECTED TO CONTROLLER/ROUTE  (" . count($directlyConnected) . " files)\n";
echo $divider . "\n";
foreach ($directlyConnected as $dotPath => $info) {
    echo "  📄 {$dotPath}\n";
    foreach ($info['referenced_in'] as $ref) {
        echo "      └─ " . basename($ref) . " ({$ref})\n";
    }
}

echo "\n";

// ── SECTION 2: Connected via blade-to-blade ──
echo $divider . "\n";
echo "🔗  CONNECTED VIA BLADE-TO-BLADE (not direct controller)  (" . count($bladeConnected) . " files)\n";
echo $divider . "\n";
foreach ($bladeConnected as $dotPath => $info) {
    echo "  📄 {$dotPath}\n";
    foreach ($info['included_by'] as $parent) {
        $parentConnected = isset($allDirectViews[$parent]) ? "✅ controller-connected" : "🔗 blade-only";
        echo "      └─ included/extended by: {$parent} [{$parentConnected}]\n";
    }
}

echo "\n";

// ── SECTION 3: NOT connected anywhere ──
echo $divider . "\n";
echo "❌  NOT CONNECTED (no controller, no blade reference)  (" . count($notConnected) . " files)\n";
echo $divider . "\n";
foreach ($notConnected as $dotPath => $info) {
    $relative = str_replace($projectRoot . '/', '', $info['path']);
    echo "  🚫 {$dotPath}\n";
    echo "      └─ File: {$relative}\n";
    if (!empty($info['included_by_blade'])) {
        foreach ($info['included_by_blade'] as $parent) {
            echo "      └─ Included by (orphan blade): {$parent}\n";
        }
    }
}

echo "\n";

// ── SECTION 4: Blade-to-blade map (for reference) ──
echo $divider . "\n";
echo "📋  BLADE-TO-BLADE REFERENCES MAP\n";
echo $divider . "\n";
foreach ($bladeToBlade as $parent => $children) {
    echo "  📄 {$parent}\n";
    foreach ($children as $child) {
        $status = isset($allBlades[$child]) ? "✅ exists" : "⚠️ FILE NOT FOUND";
        echo "      └─ @includes/extends: {$child} [{$status}]\n";
    }
}

echo "\n";
echo $divider . "\n";
echo "SUMMARY\n";
echo $divider . "\n";
echo "  ✅ Directly controller-connected : " . count($directlyConnected) . "\n";
echo "  🔗 Connected via blade-to-blade  : " . count($bladeConnected) . "\n";
echo "  ❌ Orphan / Not connected        : " . count($notConnected) . "\n";
echo "  📋 Total blade files             : " . count($allBlades) . "\n";
echo $divider . "\n";

File: project/scratch/analyze_blade.py
#!/usr/bin/env python3
"""
Blade File Connection Analyzer
Analyzes which blade files are connected to Controllers and blade-to-blade references.
"""

import os
import re
import sys
from pathlib import Path
from collections import defaultdict

PROJECT_ROOT = Path(__file__).parent.parent
VIEWS_DIR    = PROJECT_ROOT / "resources" / "views"
CONTROLLERS_DIR = PROJECT_ROOT / "app" / "Http" / "Controllers"
ROUTES_DIR   = PROJECT_ROOT / "routes"

# ─────────────────────────────────────────────────────────────
# 1. Collect all blade files -> dot.path => absolute path
# ─────────────────────────────────────────────────────────────
def get_blade_files(base_dir: Path) -> dict:
    result = {}
    for p in sorted(base_dir.rglob("*.blade.php")):
        relative = p.relative_to(base_dir)
        dot_path = str(relative).replace(os.sep, ".").replace(".blade.php", "")
        result[dot_path] = str(p)
    return result

def normalize_view_name(name: str) -> str:
    """Normalize view name: convert slashes to dots and strip whitespace.
    Laravel accepts both 'videos/create_cat' and 'videos.create_cat'."""
    return name.strip().replace('/', '.')

# ─────────────────────────────────────────────────────────────
# 2. Extract view() calls from PHP files in a directory
# ─────────────────────────────────────────────────────────────
def get_view_calls(directory: Path) -> dict:
    """Returns {normalized_dot_view_name: [file_path, ...]}"""
    view_map = defaultdict(list)
    for p in directory.rglob("*.php"):
        content = p.read_text(encoding="utf-8", errors="ignore")
        matches = re.findall(r'\bview\s*\(\s*[\'"]([^\'"]+)[\'"]', content)
        for view_name in matches:
            normalized = normalize_view_name(view_name)
            view_map[normalized].append(str(p))
    return dict(view_map)

# ─────────────────────────────────────────────────────────────
# 3. Extract blade-to-blade references (@extends, @include, @component)
# ─────────────────────────────────────────────────────────────
def get_blade_to_blade(blade_files: dict) -> dict:
    """Returns {parent_dot_path: [referenced_dot_paths]}"""
    connections = {}
    for dot_path, full_path in blade_files.items():
        content = Path(full_path).read_text(encoding="utf-8", errors="ignore")
        refs = set()

        # @extends('layout.app') or @extends('layout/app')
        for m in re.finditer(r'@extends\s*\(\s*[\'"]([^\'"]+)[\'"]', content):
            refs.add(normalize_view_name(m.group(1)))

        # @include, @includeIf, @includeWhen, @includeUnless, @includeFirst
        for m in re.finditer(r'@include(?:If|When|Unless|First)?\s*\(\s*[\'"]([^\'"]+)[\'"]', content):
            refs.add(normalize_view_name(m.group(1)))

        # @component('...')
        for m in re.finditer(r'@component\s*\(\s*[\'"]([^\'"]+)[\'"]', content):
            refs.add(normalize_view_name(m.group(1)))

        if refs:
            connections[dot_path] = sorted(refs)

    return connections

# ─────────────────────────────────────────────────────────────
# MAIN
# ─────────────────────────────────────────────────────────────
all_blades = get_blade_files(VIEWS_DIR)

# All direct view() calls from controllers & routes
controller_views = get_view_calls(CONTROLLERS_DIR)
route_views      = get_view_calls(ROUTES_DIR)

# Merged direct views
all_direct_views = {}
for view_name, files in {**controller_views, **route_views}.items():
    all_direct_views[view_name] = files

# Blade-to-blade map
blade_to_blade = get_blade_to_blade(all_blades)

# Reverse map: which blades REFERENCE this dot_path
referenced_by_blade = defaultdict(list)
for parent, children in blade_to_blade.items():
    for child in children:
        referenced_by_blade[child].append(parent)

# ─────────────────────────────────────────────────────────────
# BFS: Walk UP the include chain to find if any ancestor is
# directly controller-connected. Handles arbitrary depth, e.g.:
#   layouts.header → layouts.masterhead → dashboard (controller)
# ─────────────────────────────────────────────────────────────
def is_transitively_connected(dot_path: str, ref_by: dict, direct_views: dict):
    """
    BFS up the blade-include graph from dot_path.
    Returns True if ANY ancestor is a directly controller-connected view.
    """
    visited = {dot_path}
    queue = list(ref_by.get(dot_path, []))
    while queue:
        current = queue.pop(0)
        if current in visited:
            continue
        visited.add(current)
        if current in direct_views:
            return True
        for grandparent in ref_by.get(current, []):
            if grandparent not in visited:
                queue.append(grandparent)
    return False

# ─────────────────────────────────────────────────────────────
# Classify each blade file
# ─────────────────────────────────────────────────────────────
directly_connected = {}
blade_connected    = {}
not_connected      = {}

for dot_path, full_path in all_blades.items():
    # Check both dot-form and slash-form (Laravel accepts both)
    is_direct = dot_path in all_direct_views
    immediate_parents = referenced_by_blade.get(dot_path, [])
    rel_path = str(Path(full_path).relative_to(PROJECT_ROOT))

    if is_direct:
        all_refs = list(set(all_direct_views.get(dot_path, [])))
        ref_names = [str(Path(r).relative_to(PROJECT_ROOT)) for r in all_refs]
        directly_connected[dot_path] = {
            "file": rel_path,
            "referenced_in": sorted(ref_names),
            "raw_view_calls": sorted(set(
                k for k in (controller_views | route_views)
                if normalize_view_name(k) == dot_path
            ))
        }
    elif is_transitively_connected(dot_path, referenced_by_blade, all_direct_views):
        # BFS up full chain: handles e.g. layouts.header -> layouts.masterhead -> dashboard
        blade_connected[dot_path] = {
            "file": rel_path,
            "included_by": immediate_parents,
        }
    else:
        not_connected[dot_path] = {
            "file": rel_path,
            "included_by_blade": immediate_parents,
        }

# ─────────────────────────────────────────────────────────────
# OUTPUT
# ─────────────────────────────────────────────────────────────
DIV  = "─" * 80
DIV2 = "═" * 80

print(f"\n{'═'*78}")
print(f"{'BLADE FILE CONNECTION ANALYSIS REPORT':^78}")
print(f"{'═'*78}\n")
print(f"  Total blade files found : {len(all_blades)}")
print(f"  Views directory         : {VIEWS_DIR}")
print()

# ── Section 1: Directly connected ──────────────────────────
print(DIV)
print(f"✅  DIRECTLY CONNECTED TO CONTROLLER / ROUTE  ({len(directly_connected)} files)")
print(DIV)
for dot_path, info in sorted(directly_connected.items()):
    print(f"  📄 {dot_path}")
    for ref in info["referenced_in"]:
        raw = info.get('raw_view_calls', [])
        raw_str = f"  (called as: {', '.join(repr(r) for r in raw)})" if raw and list(raw) != [dot_path] else ""
        print(f"      └─ {ref}{raw_str}")
print()

# ── Section 2: Connected via blade chain ────────────────────
print(DIV)
print(f"🔗  CONNECTED VIA BLADE-TO-BLADE  ({len(blade_connected)} files)")
print(DIV)
for dot_path, info in sorted(blade_connected.items()):
    print(f"  📄 {dot_path}")
    for parent in info["included_by"]:
        tag = "✅ controller-connected" if parent in all_direct_views else "🔗 blade-only"
        print(f"      └─ included/extended by: {parent}  [{tag}]")
print()

# ── Section 3: Not connected ────────────────────────────────
print(DIV)
print(f"❌  NOT CONNECTED (orphan blades)  ({len(not_connected)} files)")
print(DIV)
for dot_path, info in sorted(not_connected.items()):
    print(f"  🚫 {dot_path}")
    print(f"      └─ File: {info['file']}")
    for parent in info["included_by_blade"]:
        tag = "✅ controller-connected" if parent in all_direct_views else "❌ also orphan"
        print(f"      └─ Included by (blade): {parent}  [{tag}]")
print()

# ── Section 4: Blade-to-blade map ───────────────────────────
print(DIV)
print(f"📋  BLADE-TO-BLADE REFERENCES (all @extends/@include/@component)")
print(DIV)
for parent in sorted(blade_to_blade.keys()):
    children = blade_to_blade[parent]
    print(f"  📄 {parent}")
    for child in children:
        exists = "✅ exists" if child in all_blades else "⚠️  FILE NOT FOUND IN VIEWS"
        print(f"      └─ → {child}  [{exists}]")
print()

# ── Summary ─────────────────────────────────────────────────
print(DIV)
print("SUMMARY")
print(DIV)
print(f"  ✅ Directly controller/route connected : {len(directly_connected)}")
print(f"  🔗 Blade-to-blade connected            : {len(blade_connected)}")
print(f"  ❌ Orphan / Not connected              : {len(not_connected)}")
print(f"  📋 Total blade files scanned           : {len(all_blades)}")
print(DIV)
